### PR TITLE
Contributing lib/ansible/modules/network/cloudengine/ce_stp.py module to manage HUAWEI data center CloudEngine

### PR DESCRIPTION
**Pull request by QijunPan**
##### ISSUE TYPE

 - New Module Pull Request
##### COMPONENT NAME

ce_stp.py
##### ANSIBLE VERSION

```
ansible 2.3.0
config file = /etc/ansible/ansible.cfg
configured module search path = ['/home/ansible-2.1.1.0/lib/ansible/modules/core/network/cloudengine', '/home/ansible-2.1.1.0/lib/ansible/module_utils', '/home/ansible-2.1.1.0/lib/ansible/modules', '/home/ansible-2.1.1.0', '/home/ansible-2.1.1.0/lib/ansible/utils/']
```

##### SUMMARY

```
TASK [config stp mode] *********************************************************
task path: /usr1/code/OpenNESS/code/current/KVM/intg/test/testcases/test-ce_stp.yml:23
Using module file /usr/local/lib/python2.7/site-packages/ansible-2.3.0-py2.7.egg/ansible/modules/core/network/cloudengine/ce_stp.py
<ce_6850> ESTABLISH LOCAL CONNECTION FOR USER: root
<ce_6850> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1487053900.0-25983287206973 `" && echo ansible-tmp-1487053900.0-25983287206973="` echo $HOME/.ansible/tmp/ansible-tmp-1487053900.0-25983287206973 `" ) && sleep 0'
<ce_6850> PUT /tmp/tmp1At8OQ TO /root/.ansible/tmp/ansible-tmp-1487053900.0-25983287206973/ce_stp.py
<ce_6850> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1487053900.0-25983287206973/ /root/.ansible/tmp/ansible-tmp-1487053900.0-25983287206973/ce_stp.py && sleep 0'
<ce_6850> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1487053900.0-25983287206973/ce_stp.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1487053900.0-25983287206973/" > /dev/null 2>&1 && sleep 0'
changed: [ce_6850] => {
    "changed": true, 
    "end_state": {
        "stp_mode": "stp"
    }, 
    "existing": {
        "stp_mode": "mstp"
    }, 
    "invocation": {
        "module_args": {
            "auth_pass": null, 
            "authorize": false, 
            "bpdu_filter": null, 
            "bpdu_protection": null, 
            "cost": null, 
            "edged_port": null, 
            "host": "ce_6850", 
            "interface": null, 
            "loop_protection": null, 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 12345, 
            "provider": null, 
            "root_protection": null, 
            "ssh_keyfile": null, 
            "state": "present", 
            "stp_converge": null, 
            "stp_enable": null, 
            "stp_mode": "stp", 
            "tc_protection": null, 
            "tc_protection_interval": null, 
            "tc_protection_threshold": null, 
            "timeout": 10, 
            "transport": null, 
            "use_ssl": false, 
            "username": "rootDC", 
            "validate_certs": true
        }, 
        "module_name": "ce_stp"
    }, 
    "proposed": {
        "state": "present", 
        "stp_mode": "stp"
    }, 
    "updates": [
        "stp mode stp"
    ]
}

```

### CODE REVIEW

File: lib/ansible/modules/network/cloudengine/ce_stp.py [hunk 1]
```diff
@@ -0,0 +1,969 @@
...
+    """ Module main """
+
+    argument_spec = dict(
+        state=dict(choices=['present', 'absent'], default='present'),
+        stp_mode=dict(choices=['stp', 'rstp', 'mstp']),
+        stp_enable=dict(choices=['enable', 'disable']),
+        stp_converge=dict(choices=['fast', 'normal']),
+        bpdu_protection=dict(choices=['enable', 'disable']),
+        tc_protection=dict(choices=['enable', 'disable']),
+        tc_protection_interval=dict(type='str'),
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_stp.py:953]: For integers please:
use `type='int`
remove the `isdigit` check from `check_params`

File: lib/ansible/modules/network/cloudengine/ce_stp.py [hunk 2]
```diff
@@ -0,0 +1,969 @@
...
+        tc_protection_threshold=dict(type='str'),
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_stp.py:954]: For integers please:
use `type='int`
remove the `isdigit` check from `check_params`

File: lib/ansible/modules/network/cloudengine/ce_stp.py [hunk 3]
```diff
@@ -0,0 +1,969 @@
...
+        interface=dict(type='str'),
+        edged_port=dict(choices=['enable', 'disable']),
+        bpdu_filter=dict(choices=['enable', 'disable']),
+        cost=dict(type='str'),
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_stp.py:958]: For integers please:
use `type='int`
remove the `isdigit` check from `check_params`

File: lib/ansible/modules/network/cloudengine/ce_stp.py [hunk 4]
```diff
@@ -0,0 +1,969 @@
...
+                self.module.fail_json(
+                    msg='Error: The interface %s is not exist.' % self.interface)
+
+            if "undo portswitch" in tmp_cfg:
+                self.module.fail_json(
+                    msg='Error: The interface %s is not switch mode.' % self.interface)
+
+            self.interface_stp_cfg = tmp_cfg
+
+    def check_args(self):
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_stp.py]: You've named this check_params in other modules. Not a major issue.

File: lib/ansible/modules/network/cloudengine/ce_stp.py [hunk 5]
```diff
@@ -0,0 +1,969 @@
...
+        """ Check module args """
+
+        if self.cost:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_stp.py]: Can be simplified if you do `type='int'`.

File: lib/ansible/modules/network/cloudengine/ce_stp.py [hunk 6]
```diff
@@ -0,0 +1,969 @@
...
+            if self.cost.isdigit():
+                if int(self.cost) < 1 or int(self.cost) > 200000000:
+                    self.module.fail_json(
+                        msg='Error: The value of cost is out of [1 - 200000000].')
+            else:
+                self.module.fail_json(
+                    msg='Error: The cost is not digit.')
+
+        if self.tc_protection_interval:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_stp.py:286]: Can be simplified if you do `type='int'`.

File: lib/ansible/modules/network/cloudengine/ce_stp.py [hunk 7]
```diff
@@ -0,0 +1,969 @@
...
+            if self.tc_protection_interval.isdigit():
+                if int(self.tc_protection_interval) < 1 or int(self.tc_protection_interval) > 600:
+                    self.module.fail_json(
+                        msg='Error: The value of tc_protection_interval is out of [1 - 600].')
+            else:
+                self.module.fail_json(
+                    msg='Error: The tc_protection_interval is not digit.')
+
+        if self.tc_protection_threshold:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_stp.py:295]: Can be simplified if you do `type='int'`.

File: lib/ansible/modules/network/cloudengine/ce_stp.py [hunk 8]
```diff
@@ -0,0 +1,969 @@
...
+            if self.tc_protection_threshold.isdigit():
+                if int(self.tc_protection_threshold) < 1 or int(self.tc_protection_threshold) > 255:
+                    self.module.fail_json(
+                        msg='Error: The value of tc_protection_threshold is out of [1 - 255].')
+            else:
+                self.module.fail_json(
+                    msg='Error: The tc_protection_threshold is not digit.')
+
+        if self.root_protection or self.loop_protection or self.cost:
+            if not self.interface:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_stp.py:305]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
required_together
required_one_of
require_if
```

File: lib/ansible/modules/network/cloudengine/ce_stp.py [hunk 9]
```diff
@@ -0,0 +1,969 @@
...
+                self.module.fail_json(
+                    msg='Error: Please input interface.')
+            elif self.interface == "all":
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_stp.py:308]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
mutually_exclusive
```

File: lib/ansible/modules/network/cloudengine/ce_stp.py [hunk 10]
```diff
@@ -0,0 +1,969 @@
...
+                self.module.fail_json(
+                    msg='Error: Interface can not be all when config root_protection or loop_protection or cost.')
+
+        if self.root_protection and self.root_protection == "enable":
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_stp.py:312]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
mutually_exclusive
required_together
required_one_of
require_if
```

File: lib/ansible/modules/network/cloudengine/ce_stp.py [hunk 11]
```diff
@@ -0,0 +1,969 @@
...
+            if self.loop_protection and self.loop_protection == "enable":
+                self.module.fail_json(
+                    msg='Error: Can not enable root_protection and loop_protection at the same interface.')
+
+        if self.edged_port or self.bpdu_filter:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_stp.py:317]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
required_together
require_if
```

File: lib/ansible/modules/network/cloudengine/ce_stp.py [hunk 12]
```diff
@@ -0,0 +1,969 @@
...
+        description:
+            - Set the time the MSTP device takes to handle the maximum number of TC BPDUs
+              and immediately refresh forwarding entries.
+              The value is an integer ranging from 1 to 600, in seconds.
+        required: false
+        default: null
+    tc_protection_threshold:
+        description:
+            - Set the maximum number of TC BPDUs that the MSTP can handle.
+              The value is an integer ranging from 1 to 255. The default value is 1.
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_stp.py]: I don't see `default=1` in the argspec, or in the documentation. Is this the default on the switch?

If so then this comment may need updating

File: lib/ansible/modules/network/cloudengine/ce_stp.py [hunk 13]
```diff
@@ -0,0 +1,969 @@
...
+        required: false
+        default: null
+    interface:
+        description:
+            - Interface name.
+              If the value is all, will apply configuration to all interfaces.
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_stp.py]: `If the value is C(all), apply configuration to all interfaces.`

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/cloudengine/ce_stp.py (hunk 7 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_stp.py:295]: Let's keep this type, avoid conversion

File: lib/ansible/modules/network/cloudengine/ce_stp.py (hunk 6 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_stp.py:286]: Let's keep this type, avoid conversion

File: lib/ansible/modules/network/cloudengine/ce_stp.py (hunk 3 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_stp.py:958]: Let's keep this type, avoid conversion

File: lib/ansible/modules/network/cloudengine/ce_stp.py (hunk 2 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_stp.py:954]: Let's keep this type, avoid conversion

File: lib/ansible/modules/network/cloudengine/ce_stp.py (hunk 1 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_stp.py:953]: Let's keep this type, avoid conversion

File: lib/ansible/modules/network/cloudengine/ce_stp.py (hunk 11 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_stp.py:317]: I'm trying, but it doesn't work.

File: lib/ansible/modules/network/cloudengine/ce_stp.py (hunk 8 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_stp.py:305]: I'm trying, but it doesn't work.

File: lib/ansible/modules/network/cloudengine/ce_stp.py (hunk 10 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_stp.py:312]: I'm trying, but it doesn't work.

File: lib/ansible/modules/network/cloudengine/ce_stp.py (hunk 9 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_stp.py:308]: I'm trying, but it doesn't work.

**Comment by QijunPan** [comment]
@gundalow   These problems have been fixed， can  this PR merged?

**Review by gundalow — APPROVED** [review]

**Pull request merged**